lib/internal/modules/esm/loader.js

```diff
@@ -514,9 +514,12 @@ function createModuleLoader(useCustomLoadersIfPresent = true) {
       !require('internal/modules/esm/utils').isLoaderWorker()) {
     const userLoaderPaths = getOptionValue('--experimental-loader');
     if (userLoaderPaths.length > 0) {
-      if (!emittedExperimentalWarning) {
-        emitExperimentalWarning('Custom ESM Loaders');
-        emittedExperimentalWarning = true;
+      if (!emittedLoaderFlagWarning) {
+        process.emitWarning(
+          '`--experimental-loader` may be removed in the future; instead use `--import` to reference a file that calls `register()`',
```

Comment:
I would not call out `--import` here, the beauty of `register()` is precisely that you can use it without any flag (depending on the use case)
```suggestion
          '`--experimental-loader` may be removed in the future; instead use `Module.register()`',
```

Reply:
Well but this is an error thrown _for_ a flag; without mentioning `--import` the user might not know where to put `register`. In the hooks file? As a flag `--register`? I think it’s better to be a bit more explicit. Yes there are other ways to register hooks but this is the best one to recommend because it ensures the hooks are registered before any user code runs, and `--import` can handle both CommonJS and ESM unlike `--require`.

Reply:
Well if they can't figure out what `Module.register()` is, they won't know how to use `--import` either 🤷‍♂️ What we could do it to supply a drop-in replacement in the warning, that they can copy-paste directly. wdyt?

Reply:
That would be ideal, but what would that look like? I feel like I tried to do that for the docs with `--import` and a data URL but it got really complicated with `pathToFileURL` and so on.

Reply:
Hum but actually, `globalPreload` is not called when using `register()`, so there's no "drop in replacement" 🤔 It might still be useful to give the alternate syntax using `--import`, but we need to be careful in the way we phrase it.

Reply:
On `main` the `globalPreload` hook is already gone, so I think we can land a change here that doesn’t worry about that. It’s something to consider when backporting _if_ we don’t just backport the “remove `globalPreload`“ before this.

Reply:
```suggestion
          '`--experimental-loader` may be removed in the future; instead `register()`: \n' +
          `--import 'import{register}from"node:module";import{pathToFileURL}from"node:url";${ArrayPrototypeJoin(
            ArrayPrototypeMap(userLoaderPaths, (loader) => `register(${JSONStringify(loader)}, pathToFileURL("./"))`),
            ";",
          )};`,
```

Reply:
I think you can include spaces in there, to make it more readable?

What’s an example of what this prints?

Reply:
What don't you try it out and tell me? Feel free to add spaces, but I don't know if readable is a goal

Reply:
I think being readable is a goal, this is example code we’re encouraging users to copy/paste and run.

I’ll try it out in a bit.

Reply:
So this works (linebreaks added for readability):

```shell
./node --import 'data:text/javascript,import { register } from "node:module";
  import { pathToFileURL } from "node:url";
  register("./test/fixtures/es-module-loaders/never-settling-resolve-step/loader.mjs",
  pathToFileURL("./").href);' \
  ./test/fixtures/es-module-loaders/never-settling-resolve-step/race.cjs
```

However if you neglect the `.href` it throws with `ERR_INVALID_URL`. I think we might have a bug in `register` where the second argument needs to be a string, not a string or an URL.

Reply:
> However if you neglect the `.href` it throws with `ERR_INVALID_URL`. I think we might have a bug in `register` where the second argument needs to be a string, not a string or an URL.

It seems the second argument gets ignored when it's a URL instance 😵‍💫 we should fix that ASAP